lib/utils/pagination/pagination.go

```diff
@@ -0,0 +1,118 @@
+// Teleport
+// Copyright (C) 2024 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+// Package pagination supplies vocabulary types and definitions to help make
+// paging through record sets less error prone. The goals of this package are to
+//  1. make bad conversions impossible by having separate types for the
+//     "page request" token used by the caller and "next page" token returned
+//     by the callee.
+//  2. force the single use of a PageRequestToken value, and returning an error
+//     if a token is recycled without being updated.
+//
+// For example, imagine a service that lists widgets.
+//
+//	type WidgetLister interface {
+//		ListWidgets(context.Context, *pagination.PageRequestToken) ([]Widget, NextPageToken, error)
+//	}
+//
+// Callers can page through the Widget list like so:
+//
+//	func iterateWidgets(ctx context.Context, widgetLister widgetLister) error {
+//		var pageToken pagination.PageRequestToken
+//		for {
+//			_, nextPage, err := widgetLister.ListWidgets(ctx, &pageToken)
+//			if err != nil {
+//				return trace.Wrap(err)
+//			}
+//
+//			if nextPage == pagination.EndOfList {
+//				break
+//			}
+//			pageToken.Update(nextPage)
+//		}
+//		return nil
+//	}
+//
+// A callee can retrieve the underlying token by calling Consume() on the
+// supplied PageRequestToken, which will mark the token as stale. Using a stale
+// token is an error and will cause the method to fail if a token is recycled.
+//
+//	func (_ *someType) ListWidgets(context.Context, pageRequestToken *pagination.PageRequestToken) ([]Widget, pagination.NextPageToken, error)
+//		pageToken, err := pageRequestToken.Consume()
```

Comment:
Nit: Feel free to ignore this - I'm mostly just musing on alternative solutions - and I'm sufficiently happy with the proposed solution.

Generally, I feel a little odd about having a side-effect for this. I almost wish that this consumption happened within the consumer rather than the called code. But - I see problems with that solution as well.

Reply:
What I wouldn't give for rust-like move semantics here!